Event loss due to buffer overwrites in systemMonitor.c

Steps to reproduce this issue:
Method 1: requires HostPolicy enabled
1. Run kubearmor with hostpolicy enabled.
```
vagrant@kubearmor-dev:~/KubeArmor/KubeArmor$ make && make run-with-host
```
2. Run watch command on host.(any watch command works)
```
vagrant@kubearmor-dev:~$ watch ps -au
```
3. Wait. After sometime kubearmor crashes with the following error.
```
2021-08-12 05:49:37.628234	INFO	Initialized KubeArmor
panic: runtime error: makeslice: len out of range

goroutine 69 [running]:
github.com/kubearmor/KubeArmor/KubeArmor/monitor.readByteSliceFromBuff(0x19f6f20, 0xc001532120, 0xffffffffa00b0a0a, 0x2610d58, 0x159f2e0, 0xc001ae489c, 0x0, 0x0)
	/home/vagrant/KubeArmor/KubeArmor/monitor/syscallParser.go:102 +0x72
github.com/kubearmor/KubeArmor/KubeArmor/monitor.readStringFromBuff(0x19f6f20, 0xc001532120, 0x0, 0x0, 0x0, 0x0)
	/home/vagrant/KubeArmor/KubeArmor/monitor/syscallParser.go:116 +0x169
github.com/kubearmor/KubeArmor/KubeArmor/monitor.readArgFromBuff(0x19f6f20, 0xc001532120, 0xc00118b3b0, 0x17adfa0, 0x15975a0, 0xc00118b3b0)
	/home/vagrant/KubeArmor/KubeArmor/monitor/syscallParser.go:961 +0x225
github.com/kubearmor/KubeArmor/KubeArmor/monitor.GetArgs(0xc001532120, 0xc000000002, 0x1a36c58, 0x2610d58, 0x15975a0, 0xc00118b3b0, 0x0)
	/home/vagrant/KubeArmor/KubeArmor/monitor/syscallParser.go:1038 +0x85
github.com/kubearmor/KubeArmor/KubeArmor/monitor.(*SystemMonitor).TraceSyscall(0xc0005006c0)
	/home/vagrant/KubeArmor/KubeArmor/monitor/systemMonitor.go:449 +0x338
created by github.com/kubearmor/KubeArmor/KubeArmor/core.(*KubeArmorDaemon).MonitorSystemEvents
	/home/vagrant/KubeArmor/KubeArmor/core/kubeArmor.go:238 +0x97
Makefile:37: recipe for target 'run-with-host' failed
make: *** [run-with-host] Error 2
```
Method 2: 
1. Run kubearmor.
```
vagrant@kubearmor-dev:~/KubeArmor/KubeArmor$ make && make run
```
2. Deploy multiubuntu pods.
3. Run watch command inside any pod using kubectl exec.
```
vagrant@kubearmor-dev:~$ kubectl exec -ti -n multiubuntu ubuntu-2-deployment-bc58cf44-g9jg7 -- bash
root@ubuntu-2-deployment-bc58cf44-g9jg7:/# watch ps -aux
```
4. Run watch in another pod using kubectl exec.
```
vagrant@kubearmor-dev:~$ kubectl exec -ti -n multiubuntu ubuntu-1-deployment-5d6b975744-h4cws -- bash
root@ubuntu-1-deployment-5d6b975744-h4cws:/# watch ps -aux
```
5. Wait. After sometime kubearmor crashes with the following error.
```
2021-08-12 06:00:30.411934	INFO	Initialized KubeArmor
panic: runtime error: makeslice: len out of range

goroutine 22 [running]:
github.com/kubearmor/KubeArmor/KubeArmor/monitor.readByteSliceFromBuff(0x19f6f20, 0xc0005f0900, 0xffffffff8d0b0a0a, 0x2610d58, 0x159f2e0, 0xc00187cb14, 0x0, 0x0)
	/home/vagrant/KubeArmor/KubeArmor/monitor/syscallParser.go:102 +0x72
github.com/kubearmor/KubeArmor/KubeArmor/monitor.readStringFromBuff(0x19f6f20, 0xc0005f0900, 0x0, 0x0, 0x0, 0x0)
	/home/vagrant/KubeArmor/KubeArmor/monitor/syscallParser.go:116 +0x169
github.com/kubearmor/KubeArmor/KubeArmor/monitor.readArgFromBuff(0x19f6f20, 0xc0005f0900, 0xc00220b540, 0x17adfa0, 0x15975a0, 0xc00220b540)
	/home/vagrant/KubeArmor/KubeArmor/monitor/syscallParser.go:961 +0x225
github.com/kubearmor/KubeArmor/KubeArmor/monitor.GetArgs(0xc0005f0900, 0xc000000002, 0x1a36c58, 0x2610d58, 0x15975a0, 0xc00220b540, 0x0)
	/home/vagrant/KubeArmor/KubeArmor/monitor/syscallParser.go:1038 +0x85
github.com/kubearmor/KubeArmor/KubeArmor/monitor.(*SystemMonitor).TraceSyscall(0xc000144ea0)
	/home/vagrant/KubeArmor/KubeArmor/monitor/systemMonitor.go:449 +0x338
created by github.com/kubearmor/KubeArmor/KubeArmor/core.(*KubeArmorDaemon).MonitorSystemEvents
	/home/vagrant/KubeArmor/KubeArmor/core/kubeArmor.go:238 +0x97
Makefile:27: recipe for target 'run' failed
make: *** [run] Error 2
```

--- a/KubeArmor/BPF/system_monitor.c
+++ b/KubeArmor/BPF/system_monitor.c
@@ -423,7 +423,7 @@ static __always_inline int save_str_to_buffer(bufs_t *bufs_p, void *ptr, u32 *of
         return -1;
     }
 
-    if (*off > MAX_BUFFER_SIZE - MAX_STRING_SIZE - sizeof(int)) {
+    if (*off > (MAX_BUFFER_SIZE>>1) - MAX_STRING_SIZE - sizeof(int)) {
         return 0; // not enough space - return
     }
 
@@ -445,8 +445,6 @@ static __always_inline int save_str_to_buffer(bufs_t *bufs_p, void *ptr, u32 *of
         bpf_probe_read(&(bufs_p->buf[*off & ((MAX_BUFFER_SIZE>>1) - 1)]), sizeof(int), &sz);
 
         *off += sz + sizeof(int);
-        //set_buffer_offset(*off);
-
         return sz + sizeof(int);
     }
 
@@ -462,12 +460,11 @@ static __always_inline int save_to_buffer(bufs_t *bufs_p, void *ptr, int size, u
         return 0;
     }
 
-    //u32 *off = get_buffer_offset();
     if (off == NULL) {
         return -1;
     }
 
-    if (*off > MAX_BUFFER_SIZE - MAX_ELEMENT_SIZE) {
+    if (*off > (MAX_BUFFER_SIZE>>1) - MAX_ELEMENT_SIZE) {
         return 0;
     }
 
@@ -477,13 +474,12 @@ static __always_inline int save_to_buffer(bufs_t *bufs_p, void *ptr, int size, u
 
     *off += 1;
 
-    if (*off > MAX_BUFFER_SIZE - MAX_ELEMENT_SIZE) {
+    if (*off > (MAX_BUFFER_SIZE>>1) - MAX_ELEMENT_SIZE) {
         return 0;
     }
 
     if (bpf_probe_read(&(bufs_p->buf[*off & ((MAX_BUFFER_SIZE>>1) - 1)]), size, ptr) == 0) {
         *off += size;
-        //set_buffer_offset(*off);
         return size;
     }
 
@@ -601,8 +597,8 @@ static __always_inline int events_perf_submit(struct pt_regs *ctx, u32 *base_off
         return -1;
 
     void *data = &bufs_p->buf[(*base_off & ((MAX_BUFFER_SIZE>>2) - 1))];
-    u32 size = (*off & (MAX_BUFFER_SIZE - 1)) - (*base_off & (MAX_BUFFER_SIZE - 1));
-    return sys_events.perf_submit(ctx, data, size & ((MAX_BUFFER_SIZE>>2) - 1));
+    u32 size = (*off & (MAX_BUFFER_SIZE - 1)) - (*base_off & ((MAX_BUFFER_SIZE>>2) - 1));
+    return sys_events.perf_submit(ctx, data, size & (MAX_BUF_ELEM_SIZE - 1));
 }
 
 // == Syscall Hooks (Process) == //
@@ -628,12 +624,12 @@ int syscall__execve(struct pt_regs *ctx,
         return -1;
     }
 
-    u32 base_off = (*off_global) & (MAX_BUFFER_SIZE - 1);
+    u32 base_off = (*off_global) & ((MAX_BUFFER_SIZE) - 1);
     *off_global += MAX_BUF_ELEM_SIZE;
     set_buffer_offset(*off_global);
     u32 off = base_off;
 
-    if (base_off > (MAX_BUFFER_SIZE>>2)) {
+    if (base_off >= (MAX_BUFFER_SIZE>>2)) {
         set_buffer_offset(0);
         base_off = 0;
         off = base_off;
@@ -644,9 +640,6 @@ int syscall__execve(struct pt_regs *ctx,
     bufs_t *bufs_p = get_buffer();
     if (bufs_p == NULL)
         return 0;
-    if (off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
-        return 0;
-    }
     save_context_to_buffer(bufs_p, (void*)&context, &off);
 
     save_str_to_buffer(bufs_p, (void *)filename, &off);
@@ -680,12 +673,12 @@ int trace_ret_execve(struct pt_regs *ctx)
         return -1;
     }
 
-    u32 base_off = (*off_global) & (MAX_BUFFER_SIZE - 1);
+    u32 base_off = (*off_global) & ((MAX_BUFFER_SIZE) - 1);
     *off_global += MAX_BUF_ELEM_SIZE;
     set_buffer_offset(*off_global);
     u32 off = base_off;
 
-    if (base_off > (MAX_BUFFER_SIZE>>2)) {
+    if (base_off >= (MAX_BUFFER_SIZE>>2)) {
         set_buffer_offset(0);
         base_off = 0;
         off = base_off;
@@ -696,9 +689,7 @@ int trace_ret_execve(struct pt_regs *ctx)
     bufs_t *bufs_p = get_buffer();
     if (bufs_p == NULL)
         return 0;
-    if (off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
-        return 0;
-    }
+
     save_context_to_buffer(bufs_p, (void*)&context, &off);
 
     events_perf_submit(ctx, &base_off, &off);
@@ -731,12 +722,12 @@ int syscall__execveat(struct pt_regs *ctx,
         return -1;
     }
 
-    u32 base_off = (*off_global) & (MAX_BUFFER_SIZE - 1);
+    u32 base_off = (*off_global) & ((MAX_BUFFER_SIZE) - 1);
     *off_global += MAX_BUF_ELEM_SIZE;
     set_buffer_offset(*off_global);
     u32 off = base_off;
 
-    if (base_off > (MAX_BUFFER_SIZE>>2)) {
+    if (base_off >= (MAX_BUFFER_SIZE>>2)) {
         set_buffer_offset(0);
         base_off = 0;
         off = base_off;
@@ -747,9 +738,6 @@ int syscall__execveat(struct pt_regs *ctx,
     bufs_t *bufs_p = get_buffer();
     if (bufs_p == NULL)
         return 0;
-    if (off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
-        return 0;
-    }
     save_context_to_buffer(bufs_p, (void*)&context, &off);
 
     save_to_buffer(bufs_p, (void*)&dirfd, sizeof(int), INT_T, &off);
@@ -787,12 +775,12 @@ int trace_ret_execveat(struct pt_regs *ctx)
     }
 
 
-    u32 base_off = (*off_global) & (MAX_BUFFER_SIZE - 1);
+    u32 base_off = (*off_global) & ((MAX_BUFFER_SIZE) - 1);
     *off_global += MAX_BUF_ELEM_SIZE;
     set_buffer_offset(*off_global);
     u32 off = base_off;
 
-    if (base_off > (MAX_BUFFER_SIZE>>2)) {
+    if (base_off >= (MAX_BUFFER_SIZE>>2)) {
         set_buffer_offset(0);
         base_off = 0;
         off = base_off;
@@ -803,9 +791,6 @@ int trace_ret_execveat(struct pt_regs *ctx)
     bufs_t *bufs_p = get_buffer();
     if (bufs_p == NULL)
         return 0;
-    if (off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
-        return 0;
-    }
     save_context_to_buffer(bufs_p, (void*)&context, &off);
 
     events_perf_submit(ctx, &base_off, &off);
@@ -835,12 +820,12 @@ int trace_do_exit(struct pt_regs *ctx, long code)
     }
 
 
-    u32 base_off = (*off_global) & (MAX_BUFFER_SIZE - 1);
+    u32 base_off = (*off_global) & ((MAX_BUFFER_SIZE) - 1);
     *off_global += MAX_BUF_ELEM_SIZE;
     set_buffer_offset(*off_global);
     u32 off = base_off;
 
-    if (base_off > (MAX_BUFFER_SIZE>>2)) {
+    if (base_off >= (MAX_BUFFER_SIZE>>2)) {
         set_buffer_offset(0);
         base_off = 0;
         off = base_off;
@@ -851,9 +836,6 @@ int trace_do_exit(struct pt_regs *ctx, long code)
     bufs_t *bufs_p = get_buffer();
     if (bufs_p == NULL)
         return 0;
-    if (off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
-        return 0;
-    }
     save_context_to_buffer(bufs_p, (void*)&context, &off);
 
     events_perf_submit(ctx, &base_off, &off);
@@ -955,12 +937,12 @@ static __always_inline int trace_ret_generic(u32 id, struct pt_regs *ctx, u64 ty
         return -1;
     }
 
-    u32 base_off = (*off_global) & (MAX_BUFFER_SIZE - 1);
+    u32 base_off = (*off_global) & ((MAX_BUFFER_SIZE) - 1);
     *off_global += MAX_BUF_ELEM_SIZE;
     set_buffer_offset(*off_global);
     u32 off = base_off;
 
-    if (base_off > (MAX_BUFFER_SIZE>>2)) {
+    if (base_off >= (MAX_BUFFER_SIZE>>2)) {
         set_buffer_offset(0);
         base_off = 0;
         off = base_off;
@@ -971,9 +953,6 @@ static __always_inline int trace_ret_generic(u32 id, struct pt_regs *ctx, u64 ty
     bufs_t *bufs_p = get_buffer();
     if (bufs_p == NULL)
         return 0;
-    if (off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
-        return 0;
-    }
     save_context_to_buffer(bufs_p, (void*)&context, &off);
     save_args_to_buffer(types, &args, &off);
 

--- a/KubeArmor/BPF/system_monitor.c
+++ b/KubeArmor/BPF/system_monitor.c
@@ -39,7 +39,7 @@
 // == Structures == //
 
 #define MAX_BUFFER_SIZE   32768
-#define MAX_BUF_ELEM_SIZE 10922
+#define MAX_BUF_ELEM_SIZE 256
 #define MAX_STRING_SIZE   4096
 #define MAX_STR_ARR_ELEM  20
 
@@ -592,18 +592,20 @@ static __always_inline int save_args_to_buffer(u64 types, args_t *args, u32 *off
     return 0;
 }
 
-static __always_inline int events_perf_submit(struct pt_regs *ctx, u32 base_off, u32 *off)
+static __always_inline int events_perf_submit(struct pt_regs *ctx, u32 *base_off, u32 *off)
 {
     bufs_t *bufs_p = get_buffer();
     if (bufs_p == NULL)
         return -1;
 
+    if (base_off == NULL)
+        return -1;
+
     if (off == NULL)
         return -1;
 
-    void *data = &bufs_p->buf[(base_off & ((MAX_BUFFER_SIZE>>2) - 1))];
-    u32 size = (*off & (MAX_BUFFER_SIZE - 1)) - (base_off & (MAX_BUFFER_SIZE - 1));
-    bpf_trace_printk("submitted for base_off = %u \n", base_off);
+    void *data = &bufs_p->buf[(*base_off & ((MAX_BUFFER_SIZE>>2) - 1))];
+    u32 size = (*off & (MAX_BUFFER_SIZE - 1)) - (*base_off & (MAX_BUFFER_SIZE - 1));
     return sys_events.perf_submit(ctx, data, size & ((MAX_BUFFER_SIZE>>2) - 1));
 }
 
@@ -638,7 +640,7 @@ int syscall__execve(struct pt_regs *ctx,
     if (base_off >= MAX_BUF_ELEM_SIZE)
         base_off -= MAX_BUF_ELEM_SIZE;
     off = base_off;
-    if (base_off > MAX_BUF_ELEM_SIZE) {
+    if (base_off > (MAX_BUFFER_SIZE>>2)) {
         set_buffer_offset(0);
         lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
         base_off = 0;
@@ -656,7 +658,7 @@ int syscall__execve(struct pt_regs *ctx,
     save_str_to_buffer(bufs_p, (void *)filename, &off);
     save_str_arr_to_buffer(bufs_p, __argv, &off);
 
-    events_perf_submit(ctx, base_off, &off);
+    events_perf_submit(ctx, &base_off, &off);
 
     return 0;
 }
@@ -692,7 +694,7 @@ int trace_ret_execve(struct pt_regs *ctx)
     if (base_off >= MAX_BUF_ELEM_SIZE)
         base_off -= MAX_BUF_ELEM_SIZE;
     off = base_off;
-    if (base_off > MAX_BUF_ELEM_SIZE) {
+    if (base_off > (MAX_BUFFER_SIZE>>2)) {
         set_buffer_offset(0);
         lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
         base_off = 0;
@@ -707,7 +709,7 @@ int trace_ret_execve(struct pt_regs *ctx)
     }
     save_context_to_buffer(bufs_p, (void*)&context, &off);
 
-    events_perf_submit(ctx, base_off, &off);
+    events_perf_submit(ctx, &base_off, &off);
 
     return 0;
 }
@@ -745,7 +747,7 @@ int syscall__execveat(struct pt_regs *ctx,
     if (base_off >= MAX_BUF_ELEM_SIZE)
         base_off -= MAX_BUF_ELEM_SIZE;
     off = base_off;
-    if (base_off > MAX_BUF_ELEM_SIZE) {
+    if (base_off > (MAX_BUFFER_SIZE>>2)) {
         set_buffer_offset(0);
         lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
         base_off = 0;
@@ -765,7 +767,7 @@ int syscall__execveat(struct pt_regs *ctx,
     save_str_arr_to_buffer(bufs_p, __argv, &off);
     save_to_buffer(bufs_p, (void*)&flags, sizeof(int), EXEC_FLAGS_T, &off);
 
-    events_perf_submit(ctx, base_off, &off);
+    events_perf_submit(ctx, &base_off, &off);
 
     return 0;
 }
@@ -802,7 +804,7 @@ int trace_ret_execveat(struct pt_regs *ctx)
     if (base_off >= MAX_BUF_ELEM_SIZE)
         base_off -= MAX_BUF_ELEM_SIZE;
     off = base_off;
-    if (base_off > MAX_BUF_ELEM_SIZE) {
+    if (base_off > (MAX_BUFFER_SIZE>>2)) {
         set_buffer_offset(0);
         lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
         base_off = 0;
@@ -817,7 +819,7 @@ int trace_ret_execveat(struct pt_regs *ctx)
     }
     save_context_to_buffer(bufs_p, (void*)&context, &off);
 
-    events_perf_submit(ctx, base_off, &off);
+    events_perf_submit(ctx, &base_off, &off);
 
     return 0;
 }
@@ -851,7 +853,7 @@ int trace_do_exit(struct pt_regs *ctx, long code)
     if (base_off >= MAX_BUF_ELEM_SIZE)
         base_off -= MAX_BUF_ELEM_SIZE;
     off = base_off;
-    if (base_off > MAX_BUF_ELEM_SIZE) {
+    if (base_off > (MAX_BUFFER_SIZE>>2)) {
         set_buffer_offset(0);
         lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
         base_off = 0;
@@ -866,7 +868,7 @@ int trace_do_exit(struct pt_regs *ctx, long code)
     }
     save_context_to_buffer(bufs_p, (void*)&context, &off);
 
-    events_perf_submit(ctx, base_off, &off);
+    events_perf_submit(ctx, &base_off, &off);
 
     return 0;
 }
@@ -973,7 +975,7 @@ static __always_inline int trace_ret_generic(u32 id, struct pt_regs *ctx, u64 ty
     if (base_off >= MAX_BUF_ELEM_SIZE)
         base_off -= MAX_BUF_ELEM_SIZE;
     off = base_off;
-    if (base_off > MAX_BUF_ELEM_SIZE) {
+    if (base_off > (MAX_BUFFER_SIZE>>2)) {
         set_buffer_offset(0);
         lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
         base_off = 0;
@@ -989,7 +991,7 @@ static __always_inline int trace_ret_generic(u32 id, struct pt_regs *ctx, u64 ty
     save_context_to_buffer(bufs_p, (void*)&context, &off);
     save_args_to_buffer(types, &args, &off);
 
-    events_perf_submit(ctx, base_off, &off);
+    events_perf_submit(ctx, &base_off, &off);
 
     return 0;
 }

--- a/KubeArmor/BPF/system_monitor.c
+++ b/KubeArmor/BPF/system_monitor.c
@@ -31,9 +31,15 @@
 #error Minimal required kernel version is 4.14
 #endif
 
+#ifndef lock_xadd
+# define lock_xadd(ptr, val)              \
+   ((void)__sync_fetch_and_add(ptr, val))
+#endif
+
 // == Structures == //
 
 #define MAX_BUFFER_SIZE   32768
+#define MAX_BUF_ELEM_SIZE 8192
 #define MAX_STRING_SIZE   4096
 #define MAX_STR_ARR_ELEM  20
 
@@ -397,18 +403,23 @@ static __always_inline u32* get_buffer_offset()
     return bufs_offset.lookup(&idx);
 }
 
-static __always_inline int save_context_to_buffer(bufs_t *bufs_p, void *ptr)
+static __always_inline int save_context_to_buffer(bufs_t *bufs_p, void *ptr, u32 *off)
 {
-    if (bpf_probe_read(&(bufs_p->buf[0]), sizeof(sys_context_t), ptr) == 0) {
+    if (*off > MAX_BUFFER_SIZE - sizeof(sys_context_t)) {
+        return 0; // not enough space - return
+    }
+    
+    if (bpf_probe_read(&(bufs_p->buf[*off]), sizeof(sys_context_t), ptr) == 0) {
+        *off += sizeof(sys_context_t); 
         return sizeof(sys_context_t);
     }
 
     return 0;
 }
 
-static __always_inline int save_str_to_buffer(bufs_t *bufs_p, void *ptr)
+static __always_inline int save_str_to_buffer(bufs_t *bufs_p, void *ptr, u32 *off)
 {
-    u32 *off = get_buffer_offset();
+    //u32 *off = get_buffer_offset();
     if (off == NULL) {
         return -1;
     }
@@ -435,7 +446,7 @@ static __always_inline int save_str_to_buffer(bufs_t *bufs_p, void *ptr)
         bpf_probe_read(&(bufs_p->buf[*off]), sizeof(int), &sz);
 
         *off += sz + sizeof(int);
-        set_buffer_offset(*off);
+        //set_buffer_offset(*off);
 
         return sz + sizeof(int);
     }
@@ -443,7 +454,7 @@ static __always_inline int save_str_to_buffer(bufs_t *bufs_p, void *ptr)
     return 0;
 }
 
-static __always_inline int save_to_buffer(bufs_t *bufs_p, void *ptr, int size, u8 type)
+static __always_inline int save_to_buffer(bufs_t *bufs_p, void *ptr, int size, u8 type, u32 *off)
 {
     // the biggest element that can be saved with this function should be defined here
     #define MAX_ELEMENT_SIZE sizeof(struct sockaddr_un)
@@ -452,7 +463,7 @@ static __always_inline int save_to_buffer(bufs_t *bufs_p, void *ptr, int size, u
         return 0;
     }
 
-    u32 *off = get_buffer_offset();
+    //u32 *off = get_buffer_offset();
     if (off == NULL) {
         return -1;
     }
@@ -473,46 +484,46 @@ static __always_inline int save_to_buffer(bufs_t *bufs_p, void *ptr, int size, u
 
     if (bpf_probe_read(&(bufs_p->buf[*off]), size, ptr) == 0) {
         *off += size;
-        set_buffer_offset(*off);
+        //set_buffer_offset(*off);
         return size;
     }
 
     return 0;
 }
 
-static __always_inline int save_argv(bufs_t *bufs_p, void *ptr)
+static __always_inline int save_argv(bufs_t *bufs_p, void *ptr, u32 *off)
 {
     const char *argp = NULL;
     bpf_probe_read(&argp, sizeof(argp), ptr);
 
     if (argp) {
-        return save_str_to_buffer(bufs_p, (void *)(argp));
+        return save_str_to_buffer(bufs_p, (void *)(argp), off);
     }
 
     return 0;
 }
 
-static __always_inline int save_str_arr_to_buffer(bufs_t *bufs_p, const char __user *const __user *ptr)
+static __always_inline int save_str_arr_to_buffer(bufs_t *bufs_p, const char __user *const __user *ptr, u32 *off)
 {
-    save_to_buffer(bufs_p, NULL, 0, STR_ARR_T);
+    save_to_buffer(bufs_p, NULL, 0, STR_ARR_T, off);
 
     #pragma unroll
     for (int i = 0; i < MAX_STR_ARR_ELEM; i++) {
-        if (save_argv(bufs_p, (void *)&ptr[i]) == 0) {
+        if (save_argv(bufs_p, (void *)&ptr[i], off) == 0) {
              goto out;
         }
     }
 
     char ellipsis[] = "...";
-    save_str_to_buffer(bufs_p, (void *)ellipsis);
+    save_str_to_buffer(bufs_p, (void *)ellipsis, off);
 
 out:
-    save_to_buffer(bufs_p, NULL, 0, STR_ARR_T);
+    save_to_buffer(bufs_p, NULL, 0, STR_ARR_T, off);
 
     return 0;
 }
 
-static __always_inline int save_args_to_buffer(u64 types, args_t *args)
+static __always_inline int save_args_to_buffer(u64 types, args_t *args, u32 *off)
 {
     if (types == 0) {
         return 0;
@@ -529,19 +540,19 @@ static __always_inline int save_args_to_buffer(u64 types, args_t *args)
         case NONE_T:
             break;
         case INT_T:
-            save_to_buffer(bufs_p, (void*)&(args->args[i]), sizeof(int), INT_T);
+            save_to_buffer(bufs_p, (void*)&(args->args[i]), sizeof(int), INT_T, off);
             break;
         case OPEN_FLAGS_T:
-            save_to_buffer(bufs_p, (void*)&(args->args[i]), sizeof(int), OPEN_FLAGS_T);
+            save_to_buffer(bufs_p, (void*)&(args->args[i]), sizeof(int), OPEN_FLAGS_T, off);
             break;
         case STR_T:
-            save_str_to_buffer(bufs_p, (void *)args->args[i]);
+            save_str_to_buffer(bufs_p, (void *)args->args[i], off);
             break;
         case SOCK_DOM_T:
-            save_to_buffer(bufs_p, (void*)&(args->args[i]), sizeof(int), SOCK_DOM_T);
+            save_to_buffer(bufs_p, (void*)&(args->args[i]), sizeof(int), SOCK_DOM_T, off);
             break;
         case SOCK_TYPE_T:
-            save_to_buffer(bufs_p, (void*)&(args->args[i]), sizeof(int), SOCK_TYPE_T);
+            save_to_buffer(bufs_p, (void*)&(args->args[i]), sizeof(int), SOCK_TYPE_T, off);
             break;
         case SOCKADDR_T:
             if (args->args[i]) {
@@ -549,16 +560,16 @@ static __always_inline int save_args_to_buffer(u64 types, args_t *args)
                 bpf_probe_read(&family, sizeof(short), (void*)args->args[i]);
                 switch (family) {
                 case AF_UNIX:
-                    save_to_buffer(bufs_p, (void*)(args->args[i]), sizeof(struct sockaddr_un), SOCKADDR_T);
+                    save_to_buffer(bufs_p, (void*)(args->args[i]), sizeof(struct sockaddr_un), SOCKADDR_T, off);
                     break;
                 case AF_INET:
-                    save_to_buffer(bufs_p, (void*)(args->args[i]), sizeof(struct sockaddr_in), SOCKADDR_T);
+                    save_to_buffer(bufs_p, (void*)(args->args[i]), sizeof(struct sockaddr_in), SOCKADDR_T, off);
                     break;
                 case AF_INET6:
-                    save_to_buffer(bufs_p, (void*)(args->args[i]), sizeof(struct sockaddr_in6), SOCKADDR_T);
+                    save_to_buffer(bufs_p, (void*)(args->args[i]), sizeof(struct sockaddr_in6), SOCKADDR_T, off);
                     break;
                 default:
-                    save_to_buffer(bufs_p, (void*)&family, sizeof(short), SOCKADDR_T);
+                    save_to_buffer(bufs_p, (void*)&family, sizeof(short), SOCKADDR_T, off);
                 }
             }
             break;
@@ -568,18 +579,18 @@ static __always_inline int save_args_to_buffer(u64 types, args_t *args)
     return 0;
 }
 
-static __always_inline int events_perf_submit(struct pt_regs *ctx)
+static __always_inline int events_perf_submit(struct pt_regs *ctx, u32 *off, u32 *base_off)
 {
     bufs_t *bufs_p = get_buffer();
     if (bufs_p == NULL)
         return -1;
 
-    u32 *off = get_buffer_offset();
-    if (off == NULL)
-        return -1;
+    //u32 *off = get_buffer_offset();
+    //if (off == NULL)
+    //    return -1;
 
-    void *data = bufs_p->buf;
-    int size = *off & (MAX_BUFFER_SIZE-1);
+    void *data = &(bufs_p->buf[*base_off]);
+    int size = (*off) - (*base_off);
 
     return sys_events.perf_submit(ctx, data, size);
 }
@@ -602,18 +613,37 @@ int syscall__execve(struct pt_regs *ctx,
     context.argnum = 2;
     context.retval = 0;
 
-    set_buffer_offset(sizeof(sys_context_t));
+    u32 *off_global = get_buffer_offset();
+    if (off_global == NULL) {
+        return -1;
+    }
+
+    lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
+    //set_buffer_offset(sizeof(sys_context_t));
+
+    u32 base_off = (*off_global) - MAX_BUF_ELEM_SIZE;
+    u32 off = base_off;
+    if (base_off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
+        set_buffer_offset(0);
+        lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
+        base_off = (*off_global) - MAX_BUF_ELEM_SIZE;
+        off = base_off;
+    }
+    
 
     bufs_t *bufs_p = get_buffer();
     if (bufs_p == NULL)
         return 0;
+    if (off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
+        return 0;
+    }
+    bufs_p->buf[off + (MAX_BUF_ELEM_SIZE - 1)] = 0;
+    save_context_to_buffer(bufs_p, (void*)&context, &off);
 
-    save_context_to_buffer(bufs_p, (void*)&context);
+    save_str_to_buffer(bufs_p, (void *)filename, &off);
+    save_str_arr_to_buffer(bufs_p, __argv, &off);
 
-    save_str_to_buffer(bufs_p, (void *)filename);
-    save_str_arr_to_buffer(bufs_p, __argv);
-
-    events_perf_submit(ctx);
+    events_perf_submit(ctx, &off, &base_off);
 
     return 0;
 }
@@ -636,15 +666,33 @@ int trace_ret_execve(struct pt_regs *ctx)
         return 0;
     }
 
-    set_buffer_offset(sizeof(sys_context_t));
+    u32 *off_global = get_buffer_offset();
+    if (off_global == NULL) {
+        return -1;
+    }
+
+    lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
+    //set_buffer_offset(sizeof(sys_context_t));
+
+    u32 base_off = (*off_global) - MAX_BUF_ELEM_SIZE;
+    u32 off = base_off;
+    if (base_off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
+        set_buffer_offset(0);
+        lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
+        base_off = (*off_global) - MAX_BUF_ELEM_SIZE;
+        off = base_off;
+    }
 
     bufs_t *bufs_p = get_buffer();
     if (bufs_p == NULL)
         return 0;
+    if (off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
+        return 0;
+    }
+    bufs_p->buf[off + (MAX_BUF_ELEM_SIZE - 1)] = 0;
+    save_context_to_buffer(bufs_p, (void*)&context, &off);
 
-    save_context_to_buffer(bufs_p, (void*)&context);
-
-    events_perf_submit(ctx);
+    events_perf_submit(ctx, &off, &base_off);
 
     return 0;
 }
@@ -667,20 +715,40 @@ int syscall__execveat(struct pt_regs *ctx,
     context.argnum = 4;
     context.retval = 0;
 
-    set_buffer_offset(sizeof(sys_context_t));
+    //set_buffer_offset(sizeof(sys_context_t));
+
+    u32 *off_global = get_buffer_offset();
+    if (off_global == NULL) {
+        return -1;
+    }
+
+    lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
+    //set_buffer_offset(sizeof(sys_context_t));
+
+    u32 base_off = (*off_global) - MAX_BUF_ELEM_SIZE;
+    u32 off = base_off;
+    if (base_off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
+        set_buffer_offset(0);
+        lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
+        base_off = (*off_global) - MAX_BUF_ELEM_SIZE;
+        off = base_off;
+    }
 
     bufs_t *bufs_p = get_buffer();
     if (bufs_p == NULL)
         return 0;
+    if (off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
+        return 0;
+    }
+    bufs_p->buf[off + (MAX_BUF_ELEM_SIZE - 1)] = 0;
+    save_context_to_buffer(bufs_p, (void*)&context, &off);
 
-    save_context_to_buffer(bufs_p, (void*)&context);
+    save_to_buffer(bufs_p, (void*)&dirfd, sizeof(int), INT_T, &off);
+    save_str_to_buffer(bufs_p, (void *)pathname, &off);
+    save_str_arr_to_buffer(bufs_p, __argv, &off);
+    save_to_buffer(bufs_p, (void*)&flags, sizeof(int), EXEC_FLAGS_T, &off);
 
-    save_to_buffer(bufs_p, (void*)&dirfd, sizeof(int), INT_T);
-    save_str_to_buffer(bufs_p, (void *)pathname);
-    save_str_arr_to_buffer(bufs_p, __argv);
-    save_to_buffer(bufs_p, (void*)&flags, sizeof(int), EXEC_FLAGS_T);
-
-    events_perf_submit(ctx);
+    events_perf_submit(ctx, &off, &base_off);
 
     return 0;
 }
@@ -703,15 +771,34 @@ int trace_ret_execveat(struct pt_regs *ctx)
         return 0;
     }
 
-    set_buffer_offset(sizeof(sys_context_t));
+
+    u32 *off_global = get_buffer_offset();
+    if (off_global == NULL) {
+        return -1;
+    }
+
+    lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
+    //set_buffer_offset(sizeof(sys_context_t));
+
+    u32 base_off = (*off_global) - MAX_BUF_ELEM_SIZE;
+    u32 off = base_off;
+    if (base_off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
+        set_buffer_offset(0);
+        lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
+        base_off = (*off_global) - MAX_BUF_ELEM_SIZE;
+        off = base_off;
+    }
 
     bufs_t *bufs_p = get_buffer();
     if (bufs_p == NULL)
         return 0;
+    if (off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
+        return 0;
+    }
+    bufs_p->buf[off + (MAX_BUF_ELEM_SIZE - 1)] = 0;
+    save_context_to_buffer(bufs_p, (void*)&context, &off);
 
-    save_context_to_buffer(bufs_p, (void*)&context);
-
-    events_perf_submit(ctx);
+    events_perf_submit(ctx, &off, &base_off);
 
     return 0;
 }
@@ -731,15 +818,34 @@ int trace_do_exit(struct pt_regs *ctx, long code)
 
     remove_pid_ns();
 
-    set_buffer_offset(sizeof(sys_context_t));
+
+    u32 *off_global = get_buffer_offset();
+    if (off_global == NULL) {
+        return -1;
+    }
+
+    lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
+    //set_buffer_offset(sizeof(sys_context_t));
+
+    u32 base_off = (*off_global) - MAX_BUF_ELEM_SIZE;
+    u32 off = base_off;
+    if (base_off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
+        set_buffer_offset(0);
+        lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
+        base_off = (*off_global) - MAX_BUF_ELEM_SIZE;
+        off = base_off;
+    }
 
     bufs_t *bufs_p = get_buffer();
     if (bufs_p == NULL)
         return 0;
+    if (off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
+        return 0;
+    }
+    bufs_p->buf[off + (MAX_BUF_ELEM_SIZE - 1)] = 0;
+    save_context_to_buffer(bufs_p, (void*)&context, &off);
 
-    save_context_to_buffer(bufs_p, (void*)&context);
-
-    events_perf_submit(ctx);
+    events_perf_submit(ctx, &off, &base_off);
 
     return 0;
 }
@@ -832,16 +938,35 @@ static __always_inline int trace_ret_generic(u32 id, struct pt_regs *ctx, u64 ty
         return 0;
     }
 
-    set_buffer_offset(sizeof(sys_context_t));
+
+    u32 *off_global = get_buffer_offset();
+    if (off_global == NULL) {
+        return -1;
+    }
+
+    lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
+    //set_buffer_offset(sizeof(sys_context_t));
+
+    u32 base_off = (*off_global) - MAX_BUF_ELEM_SIZE;
+    u32 off = base_off;
+    if (base_off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
+        set_buffer_offset(0);
+        lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
+        base_off = (*off_global) - MAX_BUF_ELEM_SIZE;
+        off = base_off;
+    }
 
     bufs_t *bufs_p = get_buffer();
     if (bufs_p == NULL)
         return 0;
+    if (off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
+        return 0;
+    }
+    bufs_p->buf[off + (MAX_BUF_ELEM_SIZE - 1)] = 0;
+    save_context_to_buffer(bufs_p, (void*)&context, &off);
+    save_args_to_buffer(types, &args, &off);
 
-    save_context_to_buffer(bufs_p, (void*)&context);
-    save_args_to_buffer(types, &args);
-
-    events_perf_submit(ctx);
+    events_perf_submit(ctx, &off, &base_off);
 
     return 0;
 }

--- a/KubeArmor/BPF/system_monitor.c
+++ b/KubeArmor/BPF/system_monitor.c
@@ -34,7 +34,7 @@
 // == Structures == //
 
 #define MAX_BUFFER_SIZE   32768
-#define MAX_BUF_ELEM_SIZE 256
+#define MAX_BUF_ELEM_SIZE 2048
 #define MAX_STRING_SIZE   4096
 #define MAX_STR_ARR_ELEM  20
 

--- a/KubeArmor/BPF/system_monitor.c
+++ b/KubeArmor/BPF/system_monitor.c
@@ -631,6 +631,9 @@ int syscall__execve(struct pt_regs *ctx,
 
     if (base_off >= (MAX_BUFFER_SIZE>>2)) {
         set_buffer_offset(0);
+        if (bpf_get_prandom_u32() % 2) {
+            return 0;
+        }
         base_off = 0;
         off = base_off;
         *off_global += MAX_BUF_ELEM_SIZE;
@@ -680,6 +683,9 @@ int trace_ret_execve(struct pt_regs *ctx)
 
     if (base_off >= (MAX_BUFFER_SIZE>>2)) {
         set_buffer_offset(0);
+        if (bpf_get_prandom_u32() % 2) {
+            return 0;
+        }
         base_off = 0;
         off = base_off;
         *off_global += MAX_BUF_ELEM_SIZE;
@@ -729,6 +735,9 @@ int syscall__execveat(struct pt_regs *ctx,
 
     if (base_off >= (MAX_BUFFER_SIZE>>2)) {
         set_buffer_offset(0);
+        if (bpf_get_prandom_u32() % 2) {
+            return 0;
+        }
         base_off = 0;
         off = base_off;
         *off_global += MAX_BUF_ELEM_SIZE;
@@ -782,6 +791,9 @@ int trace_ret_execveat(struct pt_regs *ctx)
 
     if (base_off >= (MAX_BUFFER_SIZE>>2)) {
         set_buffer_offset(0);
+        if (bpf_get_prandom_u32() % 2) {
+            return 0;
+        }
         base_off = 0;
         off = base_off;
         *off_global += MAX_BUF_ELEM_SIZE;
@@ -827,6 +839,9 @@ int trace_do_exit(struct pt_regs *ctx, long code)
 
     if (base_off >= (MAX_BUFFER_SIZE>>2)) {
         set_buffer_offset(0);
+        if (bpf_get_prandom_u32() % 2) {
+            return 0;
+        }
         base_off = 0;
         off = base_off;
         *off_global += MAX_BUF_ELEM_SIZE;
@@ -944,6 +959,9 @@ static __always_inline int trace_ret_generic(u32 id, struct pt_regs *ctx, u64 ty
 
     if (base_off >= (MAX_BUFFER_SIZE>>2)) {
         set_buffer_offset(0);
+        if (bpf_get_prandom_u32() % 2) {
+            return 0;
+        }
         base_off = 0;
         off = base_off;
         *off_global += MAX_BUF_ELEM_SIZE;

--- a/KubeArmor/BPF/system_monitor.c
+++ b/KubeArmor/BPF/system_monitor.c
@@ -39,7 +39,7 @@
 // == Structures == //
 
 #define MAX_BUFFER_SIZE   32768
-#define MAX_BUF_ELEM_SIZE 8192
+#define MAX_BUF_ELEM_SIZE 10922
 #define MAX_STRING_SIZE   4096
 #define MAX_STR_ARR_ELEM  20
 
@@ -405,12 +405,15 @@ static __always_inline u32* get_buffer_offset()
 
 static __always_inline int save_context_to_buffer(bufs_t *bufs_p, void *ptr, u32 *off)
 {
-    if (*off > MAX_BUFFER_SIZE - sizeof(sys_context_t)) {
+    if (off == NULL)
+        return -1;
+
+    if (*off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
         return 0; // not enough space - return
-    }
-    
-    if (bpf_probe_read(&(bufs_p->buf[*off]), sizeof(sys_context_t), ptr) == 0) {
-        *off += sizeof(sys_context_t); 
+    } 
+
+    if (bpf_probe_read(&(bufs_p->buf[*off & ((MAX_BUFFER_SIZE>>1) - 1)]), sizeof(sys_context_t), ptr) == 0) {
+        *off += sizeof(sys_context_t);
         return sizeof(sys_context_t);
     }
 
@@ -429,7 +432,7 @@ static __always_inline int save_str_to_buffer(bufs_t *bufs_p, void *ptr, u32 *of
     }
 
     u8 type = STR_T;
-    bpf_probe_read(&(bufs_p->buf[*off & (MAX_BUFFER_SIZE-1)]), 1, &type);
+    bpf_probe_read(&(bufs_p->buf[*off & ((MAX_BUFFER_SIZE>>1)-1)]), 1, &type);
 
     *off += 1;
 
@@ -437,13 +440,13 @@ static __always_inline int save_str_to_buffer(bufs_t *bufs_p, void *ptr, u32 *of
         return 0;
     }
 
-    int sz = bpf_probe_read_str(&(bufs_p->buf[*off + sizeof(int)]), MAX_STRING_SIZE, ptr);
+    int sz = bpf_probe_read_str(&(bufs_p->buf[(*off + sizeof(int)) & ((MAX_BUFFER_SIZE>>1)-1)]), MAX_STRING_SIZE, ptr);
     if (sz > 0) {
         if (*off > MAX_BUFFER_SIZE - sizeof(int)) {
             return 0;
         }
 
-        bpf_probe_read(&(bufs_p->buf[*off]), sizeof(int), &sz);
+        bpf_probe_read(&(bufs_p->buf[*off & ((MAX_BUFFER_SIZE>>1) - 1)]), sizeof(int), &sz);
 
         *off += sz + sizeof(int);
         //set_buffer_offset(*off);
@@ -472,7 +475,7 @@ static __always_inline int save_to_buffer(bufs_t *bufs_p, void *ptr, int size, u
         return 0;
     }
 
-    if (bpf_probe_read(&(bufs_p->buf[*off]), 1, &type) != 0) {
+    if (bpf_probe_read(&(bufs_p->buf[*off & ((MAX_BUFFER_SIZE>>1) - 1)]), 1, &type) != 0) {
         return 0;
     }
 
@@ -482,7 +485,7 @@ static __always_inline int save_to_buffer(bufs_t *bufs_p, void *ptr, int size, u
         return 0;
     }
 
-    if (bpf_probe_read(&(bufs_p->buf[*off]), size, ptr) == 0) {
+    if (bpf_probe_read(&(bufs_p->buf[*off & ((MAX_BUFFER_SIZE>>1) - 1)]), size, ptr) == 0) {
         *off += size;
         //set_buffer_offset(*off);
         return size;
@@ -496,6 +499,9 @@ static __always_inline int save_argv(bufs_t *bufs_p, void *ptr, u32 *off)
     const char *argp = NULL;
     bpf_probe_read(&argp, sizeof(argp), ptr);
 
+    if (off == NULL)
+        return -1;
+
     if (argp) {
         return save_str_to_buffer(bufs_p, (void *)(argp), off);
     }
@@ -506,6 +512,10 @@ static __always_inline int save_argv(bufs_t *bufs_p, void *ptr, u32 *off)
 static __always_inline int save_str_arr_to_buffer(bufs_t *bufs_p, const char __user *const __user *ptr, u32 *off)
 {
     save_to_buffer(bufs_p, NULL, 0, STR_ARR_T, off);
+
+    if (off == NULL)
+        return -1;
+    
 
     #pragma unroll
     for (int i = 0; i < MAX_STR_ARR_ELEM; i++) {
@@ -533,6 +543,9 @@ static __always_inline int save_args_to_buffer(u64 types, args_t *args, u32 *off
     if (bufs_p == NULL) {
         return 0;
     }
+
+    if (off == NULL)
+        return -1;
 
     #pragma unroll
     for (int i = 0; i < MAX_ARGS; i++) {
@@ -579,20 +592,19 @@ static __always_inline int save_args_to_buffer(u64 types, args_t *args, u32 *off
     return 0;
 }
 
-static __always_inline int events_perf_submit(struct pt_regs *ctx, u32 *off, u32 *base_off)
+static __always_inline int events_perf_submit(struct pt_regs *ctx, u32 base_off, u32 *off)
 {
     bufs_t *bufs_p = get_buffer();
     if (bufs_p == NULL)
         return -1;
 
-    //u32 *off = get_buffer_offset();
-    //if (off == NULL)
-    //    return -1;
+    if (off == NULL)
+        return -1;
 
-    void *data = &(bufs_p->buf[*base_off]);
-    int size = (*off) - (*base_off);
-
-    return sys_events.perf_submit(ctx, data, size);
+    void *data = &bufs_p->buf[(base_off & ((MAX_BUFFER_SIZE>>2) - 1))];
+    u32 size = (*off & (MAX_BUFFER_SIZE - 1)) - (base_off & (MAX_BUFFER_SIZE - 1));
+    bpf_trace_printk("submitted for base_off = %u \n", base_off);
+    return sys_events.perf_submit(ctx, data, size & ((MAX_BUFFER_SIZE>>2) - 1));
 }
 
 // == Syscall Hooks (Process) == //
@@ -621,29 +633,30 @@ int syscall__execve(struct pt_regs *ctx,
     lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
     //set_buffer_offset(sizeof(sys_context_t));
 
-    u32 base_off = (*off_global) - MAX_BUF_ELEM_SIZE;
-    u32 off = base_off;
-    if (base_off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
+    u32 base_off = (*off_global) & (MAX_BUFFER_SIZE - 1);
+    u32 off = 0;
+    if (base_off >= MAX_BUF_ELEM_SIZE)
+        base_off -= MAX_BUF_ELEM_SIZE;
+    off = base_off;
+    if (base_off > MAX_BUF_ELEM_SIZE) {
         set_buffer_offset(0);
         lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
-        base_off = (*off_global) - MAX_BUF_ELEM_SIZE;
+        base_off = 0;
         off = base_off;
     }
     
-
     bufs_t *bufs_p = get_buffer();
     if (bufs_p == NULL)
         return 0;
     if (off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
         return 0;
     }
-    bufs_p->buf[off + (MAX_BUF_ELEM_SIZE - 1)] = 0;
     save_context_to_buffer(bufs_p, (void*)&context, &off);
 
     save_str_to_buffer(bufs_p, (void *)filename, &off);
     save_str_arr_to_buffer(bufs_p, __argv, &off);
 
-    events_perf_submit(ctx, &off, &base_off);
+    events_perf_submit(ctx, base_off, &off);
 
     return 0;
 }
@@ -674,12 +687,15 @@ int trace_ret_execve(struct pt_regs *ctx)
     lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
     //set_buffer_offset(sizeof(sys_context_t));
 
-    u32 base_off = (*off_global) - MAX_BUF_ELEM_SIZE;
-    u32 off = base_off;
-    if (base_off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
+    u32 base_off = (*off_global) & (MAX_BUFFER_SIZE - 1);
+    u32 off = 0;
+    if (base_off >= MAX_BUF_ELEM_SIZE)
+        base_off -= MAX_BUF_ELEM_SIZE;
+    off = base_off;
+    if (base_off > MAX_BUF_ELEM_SIZE) {
         set_buffer_offset(0);
         lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
-        base_off = (*off_global) - MAX_BUF_ELEM_SIZE;
+        base_off = 0;
         off = base_off;
     }
 
@@ -689,10 +705,9 @@ int trace_ret_execve(struct pt_regs *ctx)
     if (off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
         return 0;
     }
-    bufs_p->buf[off + (MAX_BUF_ELEM_SIZE - 1)] = 0;
     save_context_to_buffer(bufs_p, (void*)&context, &off);
 
-    events_perf_submit(ctx, &off, &base_off);
+    events_perf_submit(ctx, base_off, &off);
 
     return 0;
 }
@@ -725,12 +740,15 @@ int syscall__execveat(struct pt_regs *ctx,
     lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
     //set_buffer_offset(sizeof(sys_context_t));
 
-    u32 base_off = (*off_global) - MAX_BUF_ELEM_SIZE;
-    u32 off = base_off;
-    if (base_off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
+    u32 base_off = (*off_global) & (MAX_BUFFER_SIZE - 1);
+    u32 off = 0;
+    if (base_off >= MAX_BUF_ELEM_SIZE)
+        base_off -= MAX_BUF_ELEM_SIZE;
+    off = base_off;
+    if (base_off > MAX_BUF_ELEM_SIZE) {
         set_buffer_offset(0);
         lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
-        base_off = (*off_global) - MAX_BUF_ELEM_SIZE;
+        base_off = 0;
         off = base_off;
     }
 
@@ -740,7 +758,6 @@ int syscall__execveat(struct pt_regs *ctx,
     if (off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
         return 0;
     }
-    bufs_p->buf[off + (MAX_BUF_ELEM_SIZE - 1)] = 0;
     save_context_to_buffer(bufs_p, (void*)&context, &off);
 
     save_to_buffer(bufs_p, (void*)&dirfd, sizeof(int), INT_T, &off);
@@ -748,7 +765,7 @@ int syscall__execveat(struct pt_regs *ctx,
     save_str_arr_to_buffer(bufs_p, __argv, &off);
     save_to_buffer(bufs_p, (void*)&flags, sizeof(int), EXEC_FLAGS_T, &off);
 
-    events_perf_submit(ctx, &off, &base_off);
+    events_perf_submit(ctx, base_off, &off);
 
     return 0;
 }
@@ -780,12 +797,15 @@ int trace_ret_execveat(struct pt_regs *ctx)
     lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
     //set_buffer_offset(sizeof(sys_context_t));
 
-    u32 base_off = (*off_global) - MAX_BUF_ELEM_SIZE;
-    u32 off = base_off;
-    if (base_off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
+    u32 base_off = (*off_global) & (MAX_BUFFER_SIZE - 1);
+    u32 off = 0;
+    if (base_off >= MAX_BUF_ELEM_SIZE)
+        base_off -= MAX_BUF_ELEM_SIZE;
+    off = base_off;
+    if (base_off > MAX_BUF_ELEM_SIZE) {
         set_buffer_offset(0);
         lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
-        base_off = (*off_global) - MAX_BUF_ELEM_SIZE;
+        base_off = 0;
         off = base_off;
     }
 
@@ -795,10 +815,9 @@ int trace_ret_execveat(struct pt_regs *ctx)
     if (off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
         return 0;
     }
-    bufs_p->buf[off + (MAX_BUF_ELEM_SIZE - 1)] = 0;
     save_context_to_buffer(bufs_p, (void*)&context, &off);
 
-    events_perf_submit(ctx, &off, &base_off);
+    events_perf_submit(ctx, base_off, &off);
 
     return 0;
 }
@@ -827,12 +846,15 @@ int trace_do_exit(struct pt_regs *ctx, long code)
     lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
     //set_buffer_offset(sizeof(sys_context_t));
 
-    u32 base_off = (*off_global) - MAX_BUF_ELEM_SIZE;
-    u32 off = base_off;
-    if (base_off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
+    u32 base_off = (*off_global) & (MAX_BUFFER_SIZE - 1);
+    u32 off = 0;
+    if (base_off >= MAX_BUF_ELEM_SIZE)
+        base_off -= MAX_BUF_ELEM_SIZE;
+    off = base_off;
+    if (base_off > MAX_BUF_ELEM_SIZE) {
         set_buffer_offset(0);
         lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
-        base_off = (*off_global) - MAX_BUF_ELEM_SIZE;
+        base_off = 0;
         off = base_off;
     }
 
@@ -842,10 +864,9 @@ int trace_do_exit(struct pt_regs *ctx, long code)
     if (off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
         return 0;
     }
-    bufs_p->buf[off + (MAX_BUF_ELEM_SIZE - 1)] = 0;
     save_context_to_buffer(bufs_p, (void*)&context, &off);
 
-    events_perf_submit(ctx, &off, &base_off);
+    events_perf_submit(ctx, base_off, &off);
 
     return 0;
 }
@@ -947,12 +968,15 @@ static __always_inline int trace_ret_generic(u32 id, struct pt_regs *ctx, u64 ty
     lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
     //set_buffer_offset(sizeof(sys_context_t));
 
-    u32 base_off = (*off_global) - MAX_BUF_ELEM_SIZE;
-    u32 off = base_off;
-    if (base_off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
+    u32 base_off = (*off_global) & (MAX_BUFFER_SIZE - 1);
+    u32 off = 0;
+    if (base_off >= MAX_BUF_ELEM_SIZE)
+        base_off -= MAX_BUF_ELEM_SIZE;
+    off = base_off;
+    if (base_off > MAX_BUF_ELEM_SIZE) {
         set_buffer_offset(0);
         lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
-        base_off = (*off_global) - MAX_BUF_ELEM_SIZE;
+        base_off = 0;
         off = base_off;
     }
 
@@ -962,11 +986,10 @@ static __always_inline int trace_ret_generic(u32 id, struct pt_regs *ctx, u64 ty
     if (off > MAX_BUFFER_SIZE - MAX_BUF_ELEM_SIZE) {
         return 0;
     }
-    bufs_p->buf[off + (MAX_BUF_ELEM_SIZE - 1)] = 0;
     save_context_to_buffer(bufs_p, (void*)&context, &off);
     save_args_to_buffer(types, &args, &off);
 
-    events_perf_submit(ctx, &off, &base_off);
+    events_perf_submit(ctx, base_off, &off);
 
     return 0;
 }

--- a/KubeArmor/BPF/system_monitor.c
+++ b/KubeArmor/BPF/system_monitor.c
@@ -31,10 +31,6 @@
 #error Minimal required kernel version is 4.14
 #endif
 
-#ifndef lock_xadd
-# define lock_xadd(ptr, val)              \
-   ((void)__sync_fetch_and_add(ptr, val))
-#endif
 
 // == Structures == //
 
@@ -632,19 +628,17 @@ int syscall__execve(struct pt_regs *ctx,
         return -1;
     }
 
-    lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
-    //set_buffer_offset(sizeof(sys_context_t));
-
     u32 base_off = (*off_global) & (MAX_BUFFER_SIZE - 1);
-    u32 off = 0;
-    if (base_off >= MAX_BUF_ELEM_SIZE)
-        base_off -= MAX_BUF_ELEM_SIZE;
-    off = base_off;
+    *off_global += MAX_BUF_ELEM_SIZE;
+    set_buffer_offset(*off_global);
+    u32 off = base_off;
+
     if (base_off > (MAX_BUFFER_SIZE>>2)) {
         set_buffer_offset(0);
-        lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
         base_off = 0;
         off = base_off;
+        *off_global += MAX_BUF_ELEM_SIZE;
+        set_buffer_offset(*off_global);
     }
     
     bufs_t *bufs_p = get_buffer();
@@ -686,19 +680,17 @@ int trace_ret_execve(struct pt_regs *ctx)
         return -1;
     }
 
-    lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
-    //set_buffer_offset(sizeof(sys_context_t));
-
     u32 base_off = (*off_global) & (MAX_BUFFER_SIZE - 1);
-    u32 off = 0;
-    if (base_off >= MAX_BUF_ELEM_SIZE)
-        base_off -= MAX_BUF_ELEM_SIZE;
-    off = base_off;
+    *off_global += MAX_BUF_ELEM_SIZE;
+    set_buffer_offset(*off_global);
+    u32 off = base_off;
+
     if (base_off > (MAX_BUFFER_SIZE>>2)) {
         set_buffer_offset(0);
-        lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
         base_off = 0;
         off = base_off;
+        *off_global += MAX_BUF_ELEM_SIZE;
+        set_buffer_offset(*off_global);
     }
 
     bufs_t *bufs_p = get_buffer();
@@ -739,19 +731,17 @@ int syscall__execveat(struct pt_regs *ctx,
         return -1;
     }
 
-    lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
-    //set_buffer_offset(sizeof(sys_context_t));
-
     u32 base_off = (*off_global) & (MAX_BUFFER_SIZE - 1);
-    u32 off = 0;
-    if (base_off >= MAX_BUF_ELEM_SIZE)
-        base_off -= MAX_BUF_ELEM_SIZE;
-    off = base_off;
+    *off_global += MAX_BUF_ELEM_SIZE;
+    set_buffer_offset(*off_global);
+    u32 off = base_off;
+
     if (base_off > (MAX_BUFFER_SIZE>>2)) {
         set_buffer_offset(0);
-        lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
         base_off = 0;
         off = base_off;
+        *off_global += MAX_BUF_ELEM_SIZE;
+        set_buffer_offset(*off_global);
     }
 
     bufs_t *bufs_p = get_buffer();
@@ -796,19 +786,18 @@ int trace_ret_execveat(struct pt_regs *ctx)
         return -1;
     }
 
-    lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
-    //set_buffer_offset(sizeof(sys_context_t));
 
     u32 base_off = (*off_global) & (MAX_BUFFER_SIZE - 1);
-    u32 off = 0;
-    if (base_off >= MAX_BUF_ELEM_SIZE)
-        base_off -= MAX_BUF_ELEM_SIZE;
-    off = base_off;
+    *off_global += MAX_BUF_ELEM_SIZE;
+    set_buffer_offset(*off_global);
+    u32 off = base_off;
+
     if (base_off > (MAX_BUFFER_SIZE>>2)) {
         set_buffer_offset(0);
-        lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
         base_off = 0;
         off = base_off;
+        *off_global += MAX_BUF_ELEM_SIZE;
+        set_buffer_offset(*off_global);
     }
 
     bufs_t *bufs_p = get_buffer();
@@ -845,19 +834,18 @@ int trace_do_exit(struct pt_regs *ctx, long code)
         return -1;
     }
 
-    lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
-    //set_buffer_offset(sizeof(sys_context_t));
 
     u32 base_off = (*off_global) & (MAX_BUFFER_SIZE - 1);
-    u32 off = 0;
-    if (base_off >= MAX_BUF_ELEM_SIZE)
-        base_off -= MAX_BUF_ELEM_SIZE;
-    off = base_off;
+    *off_global += MAX_BUF_ELEM_SIZE;
+    set_buffer_offset(*off_global);
+    u32 off = base_off;
+
     if (base_off > (MAX_BUFFER_SIZE>>2)) {
         set_buffer_offset(0);
-        lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
         base_off = 0;
         off = base_off;
+        *off_global += MAX_BUF_ELEM_SIZE;
+        set_buffer_offset(*off_global);
     }
 
     bufs_t *bufs_p = get_buffer();
@@ -967,19 +955,17 @@ static __always_inline int trace_ret_generic(u32 id, struct pt_regs *ctx, u64 ty
         return -1;
     }
 
-    lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
-    //set_buffer_offset(sizeof(sys_context_t));
-
     u32 base_off = (*off_global) & (MAX_BUFFER_SIZE - 1);
-    u32 off = 0;
-    if (base_off >= MAX_BUF_ELEM_SIZE)
-        base_off -= MAX_BUF_ELEM_SIZE;
-    off = base_off;
+    *off_global += MAX_BUF_ELEM_SIZE;
+    set_buffer_offset(*off_global);
+    u32 off = base_off;
+
     if (base_off > (MAX_BUFFER_SIZE>>2)) {
         set_buffer_offset(0);
-        lock_xadd(off_global, MAX_BUF_ELEM_SIZE);
         base_off = 0;
         off = base_off;
+        *off_global += MAX_BUF_ELEM_SIZE;
+        set_buffer_offset(*off_global);
     }
 
     bufs_t *bufs_p = get_buffer();

--- a/KubeArmor/BPF/system_monitor.c
+++ b/KubeArmor/BPF/system_monitor.c
@@ -31,7 +31,6 @@
 #error Minimal required kernel version is 4.14
 #endif
 
-
 // == Structures == //
 
 #define MAX_BUFFER_SIZE   32768
@@ -418,7 +417,6 @@ static __always_inline int save_context_to_buffer(bufs_t *bufs_p, void *ptr, u32
 
 static __always_inline int save_str_to_buffer(bufs_t *bufs_p, void *ptr, u32 *off)
 {
-    //u32 *off = get_buffer_offset();
     if (off == NULL) {
         return -1;
     }
@@ -598,6 +596,7 @@ static __always_inline int events_perf_submit(struct pt_regs *ctx, u32 *base_off
 
     void *data = &bufs_p->buf[(*base_off & ((MAX_BUFFER_SIZE>>2) - 1))];
     u32 size = (*off & (MAX_BUFFER_SIZE - 1)) - (*base_off & ((MAX_BUFFER_SIZE>>2) - 1));
+
     return sys_events.perf_submit(ctx, data, size & (MAX_BUF_ELEM_SIZE - 1));
 }
 
@@ -643,6 +642,7 @@ int syscall__execve(struct pt_regs *ctx,
     bufs_t *bufs_p = get_buffer();
     if (bufs_p == NULL)
         return 0;
+
     save_context_to_buffer(bufs_p, (void*)&context, &off);
 
     save_str_to_buffer(bufs_p, (void *)filename, &off);
@@ -721,8 +721,6 @@ int syscall__execveat(struct pt_regs *ctx,
     context.argnum = 4;
     context.retval = 0;
 
-    //set_buffer_offset(sizeof(sys_context_t));
-
     u32 *off_global = get_buffer_offset();
     if (off_global == NULL) {
         return -1;
@@ -747,6 +745,7 @@ int syscall__execveat(struct pt_regs *ctx,
     bufs_t *bufs_p = get_buffer();
     if (bufs_p == NULL)
         return 0;
+
     save_context_to_buffer(bufs_p, (void*)&context, &off);
 
     save_to_buffer(bufs_p, (void*)&dirfd, sizeof(int), INT_T, &off);
@@ -777,12 +776,10 @@ int trace_ret_execveat(struct pt_regs *ctx)
         return 0;
     }
 
-
     u32 *off_global = get_buffer_offset();
     if (off_global == NULL) {
         return -1;
     }
-
 
     u32 base_off = (*off_global) & ((MAX_BUFFER_SIZE) - 1);
     *off_global += MAX_BUF_ELEM_SIZE;
@@ -825,12 +822,10 @@ int trace_do_exit(struct pt_regs *ctx, long code)
 
     remove_pid_ns();
 
-
     u32 *off_global = get_buffer_offset();
     if (off_global == NULL) {
         return -1;
     }
-
 
     u32 base_off = (*off_global) & ((MAX_BUFFER_SIZE) - 1);
     *off_global += MAX_BUF_ELEM_SIZE;
@@ -851,6 +846,7 @@ int trace_do_exit(struct pt_regs *ctx, long code)
     bufs_t *bufs_p = get_buffer();
     if (bufs_p == NULL)
         return 0;
+
     save_context_to_buffer(bufs_p, (void*)&context, &off);
 
     events_perf_submit(ctx, &base_off, &off);
@@ -971,6 +967,7 @@ static __always_inline int trace_ret_generic(u32 id, struct pt_regs *ctx, u64 ty
     bufs_t *bufs_p = get_buffer();
     if (bufs_p == NULL)
         return 0;
+
     save_context_to_buffer(bufs_p, (void*)&context, &off);
     save_args_to_buffer(types, &args, &off);
 

